src/track/beats.cpp

```diff
@@ -53,3 +190,676 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
     return nthBeat;
 };
 
+void Beats::onBeatlistChanged() {
+    if (!isValid()) {
+        m_dLastFrame = 0;
+        m_dCachedBpm = 0;
+        return;
+    }
+    m_dLastFrame = m_beats.last().frame_position();
+    track::io::Beat startBeat = m_beats.first();
+    track::io::Beat stopBeat = m_beats.last();
+    m_dCachedBpm = calculateBpm(startBeat, stopBeat);
+}
+
+bool Beats::isValid() const {
+    return m_iSampleRate > 0 && m_beats.size() > 0;
+}
+
+double Beats::calculateBpm(const track::io::Beat& startBeat, const track::io::Beat& stopBeat) const {
+    if (startBeat.frame_position() > stopBeat.frame_position()) {
+        return -1;
+    }
+
+    BeatList::const_iterator curBeat =
```

Comment:
You iterate three times over the same list. 
Can we join the loops?

Reply:
Now it works. There are other topics more urgent so I will look at this later.
I will leave a TODO for me to look at it in the future.
This works for you?